components/ContactList.js

```diff
@@ -27,18 +34,60 @@ import ContactItem from "../components/ContactItem";
 // .....
 
 export default function ContactList() {
+  // Hooks
   const dispatch = useDispatch();
   const contacts = useSelector(getContacts);
+  const favorites = useSelector(getFavorites);
+  const [showFavorites, setShowFavorites] = useState(false);
 
   useEffect(() => {
     dispatch(fetchContacts());
   }, []);
 
+  // JSX
+  const contactsJSX = contacts.map((contact, index, array) => {
+    const currentLetter = contact.lastName[0].toUpperCase();
+    const favorite = favorites.some((favorite) => favorite.id === contact.id);
+    if (!index) {
+      // Provides header for first contact item
+      return (
+        <Fragment key={index}>
+          <p>{currentLetter}</p>
+          <ContactItem key={index} {...contact} favorite={favorite} />
+        </Fragment>
+      );
+    } else if (index < array.length - 1) {
+      const nextLetter = array[index + 1].lastName[0].toUpperCase();
+      if (currentLetter !== nextLetter) {
+        // Provides header for all other contact items
+        return (
+          <Fragment key={index}>
+            <ContactItem key={index} {...contact} favorite={favorite} />
+            <p>{nextLetter}</p>
+          </Fragment>
+        )
+      } else {
+        return (
+          <ContactItem key={index} {...contact} favorite={favorite} />
+        )
+      }
+    }
+  });
+
+  const favoritesJSX = favorites.map((contact, index, array) => {
+    const favorite = favorites.some((favorite) => favorite.id === contact.id);
```

Comment:
probably unnecessary? isn't favorite the same as contact in this case?

Reply:
You are right as the elements inside the favorites list is the same as the contacts list and I was initially going to write it something like this:
```
const [list, setList] = useState([]);

useEffect(() => {
  if (showFavorites) {
    setList(favorites);
  } else {
    setList(contacts);
  }
}, [showFavorites, contacts, favorites]);

const listJSX = list.map((contact, index, array) => {
  ...
});

return (
  <div>
    {listJSX}
  </div>
)
```
However, I ran into issues when I was rendering the favorites list when switching between `all` and `favorite` contacts. The main issue was that it would render all of the favorites contacts except for the most recently favorited contact and also the letter headers sometimes did not render properly when the user decided to unfavorite a contact inside the favorites list.
Given more time I believe that a majority of these issues could have been resolved by encapsulating the contacts list JSX element into its own component where it would accept a list of contacts as one of its props. However, I decided to work around this issue by separating the two into their own JSX elements so that each would be recreated when toggling between `favorites` and `all`. For this small case, this solution seems to have worked however for lists with a lot more elements, this would impact performance as all the elements would be recreated when switching views.